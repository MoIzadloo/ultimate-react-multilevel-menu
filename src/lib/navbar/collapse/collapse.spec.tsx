import { render, fireEvent } from '@testing-library/react'
import React from 'react'
import { Navbar, Collapse, Item, Items, Logo } from '../'
import classNames from 'classnames'

const testid = 'navbar'

const children = (
    <>
        <Logo href="/">Logo</Logo>
        <Collapse>
            <Item href={'#home'}>Home</Item>
            <Items href={'#articles'} title="Articles">
                <Items href={'#power-plant'} title="Power plant">
                    <Item href={'#turbine'}>Turbine</Item>
                </Items>
            </Items>
            <Items href={'#about'} title="About">
                <Item href={'#introduction'}>Introduction</Item>
            </Items>
            <Item href={'#pricing'}>Pricing</Item>
        </Collapse>
    </>
)

describe('<Collapse>', () => {
    it('matches snapshot', () => {
        const body = renderNavbar({}) 
        expect(body).toMatchSnapshot()
    })
    it('renders correctly', () => {
        const { getByTestId } = renderNavbar({})
        const navElemnt = getByTestId(testid).getElementsByClassName("nav")[0]
        const collapseElement = getByTestId(testid).getElementsByClassName("navbar-collapse")[0]
        const togglerElement = getByTestId(testid).getElementsByClassName("navbar-toggler")[0]
        expect(togglerElement.firstElementChild?.classList.contains('navbar-toggler-line')).toBe(true)
        expect(navElemnt.classList.contains('nav')).toBe(true)
        expect(navElemnt.children.length).toBe(4)
        expect(collapseElement.classList.contains('show')).toBe(false)
        expect(collapseElement.children.length).toBe(4)
        fireEvent.click(togglerElement)
        expect(togglerElement.classList.contains('collapsed')).toBe(true)
        const firstItem = collapseElement.getElementsByClassName("item")[0]
        fireEvent.click(firstItem)
        expect(togglerElement.classList.contains('collapsed')).toBe(false)
        fireEvent.click(togglerElement)
        expect(togglerElement.classList.contains('collapsed')).toBe(true)
        const firstItems = collapseElement.getElementsByClassName("dropdown")[0]
        fireEvent.click(firstItems)
        expect(togglerElement.classList.contains('collapsed')).toBe(true)
        fireEvent.click(firstItems)
        expect(togglerElement.classList.contains('collapsed')).toBe(false)
    })
    it('test toggler button', () =>{
        const { getByTestId } = renderNavbar({})
        const togglerElement = getByTestId(testid).getElementsByClassName("navbar-toggler")[0]
        fireEvent.click(togglerElement)
        expect(togglerElement.classList.contains('collapsed')).toBe(true)
        fireEvent.click(togglerElement)
        expect(togglerElement.classList.contains('collapsed')).toBe(false)
    })
    it('should render children', () => {
        const { getByTestId } = renderNavbar({})
        const navElemnt = getByTestId(testid).getElementsByClassName("nav")[0]
        expect(navElemnt.children.length).toBeGreaterThan(0)
    })
})

const defaultProps = {
    children,
    'data-testid': testid,
    className: classNames('navbar-light', 'bg-white')
}

const renderNavbar = ({...props}) =>{
    return render(
    <Navbar 
        {...defaultProps}
        {...props}
    />)
}