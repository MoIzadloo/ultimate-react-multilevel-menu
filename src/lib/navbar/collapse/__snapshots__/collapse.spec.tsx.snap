// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Collapse> matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav
        class="navbar navbar-light bg-white"
        data-testid="navbar"
      >
        <div
          class="logo-container"
        >
          <a
            href="/"
          >
            Logo
          </a>
        </div>
        <div
          class="navbar-toggler"
        >
          <span
            class="navbar-toggler-line"
          />
          <span
            class="navbar-toggler-line"
          />
          <span
            class="navbar-toggler-line"
          />
        </div>
        <ul
          class="nav"
        >
          <li
            class="nav-item"
          >
            <a
              class="item"
              href="#home"
            >
              Home
            </a>
          </li>
          <li
            class="nav-item is-navbar"
          >
            <a
              class="item dropdown"
              href="#articles"
            >
              Articles
            </a>
            <ul
              class="nav-items"
            >
              <li
                class="nav-item is-navbar"
              >
                <a
                  class="item dropdown"
                  href="#power-plant"
                >
                  Power plant
                </a>
                <ul
                  class="nav-items"
                >
                  <li
                    class="nav-item"
                  >
                    <a
                      class="item"
                      href="#turbine"
                    >
                      Turbine
                    </a>
                  </li>
                </ul>
              </li>
            </ul>
          </li>
          <li
            class="nav-item is-navbar"
          >
            <a
              class="item dropdown"
              href="#about"
            >
              About
            </a>
            <ul
              class="nav-items"
            >
              <li
                class="nav-item"
              >
                <a
                  class="item"
                  href="#introduction"
                >
                  Introduction
                </a>
              </li>
            </ul>
          </li>
          <li
            class="nav-item"
          >
            <a
              class="item"
              href="#pricing"
            >
              Pricing
            </a>
          </li>
        </ul>
        <ul
          class="navbar-collapse"
        >
          <li
            class="nav-item"
          >
            <a
              class="item"
              href="#home"
            >
              Home
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              class="item dropdown"
              href="#articles"
            >
              Articles
            </a>
            <ul
              class="nav-items"
            >
              <li
                class="nav-item"
              >
                <a
                  class="item dropdown"
                  href="#power-plant"
                >
                  Power plant
                </a>
                <ul
                  class="nav-items"
                >
                  <li
                    class="nav-item"
                  >
                    <a
                      class="item"
                      href="#turbine"
                    >
                      Turbine
                    </a>
                  </li>
                </ul>
              </li>
            </ul>
          </li>
          <li
            class="nav-item"
          >
            <a
              class="item dropdown"
              href="#about"
            >
              About
            </a>
            <ul
              class="nav-items"
            >
              <li
                class="nav-item"
              >
                <a
                  class="item"
                  href="#introduction"
                >
                  Introduction
                </a>
              </li>
            </ul>
          </li>
          <li
            class="nav-item"
          >
            <a
              class="item"
              href="#pricing"
            >
              Pricing
            </a>
          </li>
        </ul>
      </nav>
    </div>
  </body>,
  "container": <div>
    <nav
      class="navbar navbar-light bg-white"
      data-testid="navbar"
    >
      <div
        class="logo-container"
      >
        <a
          href="/"
        >
          Logo
        </a>
      </div>
      <div
        class="navbar-toggler"
      >
        <span
          class="navbar-toggler-line"
        />
        <span
          class="navbar-toggler-line"
        />
        <span
          class="navbar-toggler-line"
        />
      </div>
      <ul
        class="nav"
      >
        <li
          class="nav-item"
        >
          <a
            class="item"
            href="#home"
          >
            Home
          </a>
        </li>
        <li
          class="nav-item is-navbar"
        >
          <a
            class="item dropdown"
            href="#articles"
          >
            Articles
          </a>
          <ul
            class="nav-items"
          >
            <li
              class="nav-item is-navbar"
            >
              <a
                class="item dropdown"
                href="#power-plant"
              >
                Power plant
              </a>
              <ul
                class="nav-items"
              >
                <li
                  class="nav-item"
                >
                  <a
                    class="item"
                    href="#turbine"
                  >
                    Turbine
                  </a>
                </li>
              </ul>
            </li>
          </ul>
        </li>
        <li
          class="nav-item is-navbar"
        >
          <a
            class="item dropdown"
            href="#about"
          >
            About
          </a>
          <ul
            class="nav-items"
          >
            <li
              class="nav-item"
            >
              <a
                class="item"
                href="#introduction"
              >
                Introduction
              </a>
            </li>
          </ul>
        </li>
        <li
          class="nav-item"
        >
          <a
            class="item"
            href="#pricing"
          >
            Pricing
          </a>
        </li>
      </ul>
      <ul
        class="navbar-collapse"
      >
        <li
          class="nav-item"
        >
          <a
            class="item"
            href="#home"
          >
            Home
          </a>
        </li>
        <li
          class="nav-item"
        >
          <a
            class="item dropdown"
            href="#articles"
          >
            Articles
          </a>
          <ul
            class="nav-items"
          >
            <li
              class="nav-item"
            >
              <a
                class="item dropdown"
                href="#power-plant"
              >
                Power plant
              </a>
              <ul
                class="nav-items"
              >
                <li
                  class="nav-item"
                >
                  <a
                    class="item"
                    href="#turbine"
                  >
                    Turbine
                  </a>
                </li>
              </ul>
            </li>
          </ul>
        </li>
        <li
          class="nav-item"
        >
          <a
            class="item dropdown"
            href="#about"
          >
            About
          </a>
          <ul
            class="nav-items"
          >
            <li
              class="nav-item"
            >
              <a
                class="item"
                href="#introduction"
              >
                Introduction
              </a>
            </li>
          </ul>
        </li>
        <li
          class="nav-item"
        >
          <a
            class="item"
            href="#pricing"
          >
            Pricing
          </a>
        </li>
      </ul>
    </nav>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
